.github/workflows/py-generate.yml

```diff
@@ -0,0 +1,23 @@
+name: Generate Pydantic Models
+
+on:
+  push:
+    branches:
+      - main
+    paths:
+      - 'catalog-rest-service/src/main/resources/json/**'
+
+jobs:
+  report:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
```

Comment:
When does this action run ?

While CI checks out main branch?

Reply:
Hi, @mithmatt,

At the top of the action we have defined
```
on:
  push:
    branches:
      - main
    paths:
      - 'catalog-rest-service/src/main/resources/json/**'
```

So it will run after a push to `main` if there are updated files under the JSON schemas dir. This way we can be sure that the origin's pydantic models are up to date with any update on the JSON schemas.

Reply:
Thanks for clarifying that. 

If I understand correctly, following are the sequence of events:

1. Developer changes json schema, writes and commits code (commit A)
2. Committer merges code into main branch
3. Githhub workflow kicks in
4. Github creates another commit (commit B) with pydantic models


Should pydantic models be generated as part of commit A itself, so that they go in as one single commit?
I think a client side git [pre-commit](https://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks) hook is more appropriate for this use case.

Reply:
Yep exactly, that is the workflow.

I also thought about coupling/decoupling the commits and while it might feel cleaner to have it in a single shot, I also believe it is important that the process acts in an independent and reproducible way. As we want to encourage different collaborators to play around with the schemas, adding an extra layer of complexity where they need to have the right environment and packages ready might be a bit too much.

I feel it might be more valuable to have a simple process that we can easily track even if that means having a specific commit where the codebase won't be aligned.

WDYT?

> Also that might me remember that I need to install the python dependencies in the action!

Reply:
I see we are trading off correctness for simplicity + developer velocity with the github workflow. 
I agree it is better to have a simple process.

For jsonschema to java pojos - mvn generate-sources (runs as part of mvn test) ensures that the classes can be generated successfully when a pull request is raised.

Should we have a similar one for pydantic models (to prevent us from checking in code where this generation could potentially fail)?

PS: I'm probably extra cautious about breaking something.

Reply:
actually, you have a good point, as not that long ago we needed to slightly update a JSON schema change as the pydantic generation failed.

As part of the CI process, we could have a separate job that validates this conversion, even if it does not update the current PR. I also would advise against validating and generating the sources during the PR as the review becomes quite cumbersome, with a lot of changes we do not even want to see.

But yep, +1 on being extra cautious here! I can prepare a similar workflow to run with each PR. Let's try to keep the automation as safe as possible.